Trellix/client/http_client.py

```diff
@@ -0,0 +1,242 @@
+
+    @property
+    def _edr_historical_searches_query_url(self) -> str:
+        """
+        Get request url for historical searches query.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/searches/historical".format(self.base_url)
+
+    @property
+    def _edr_investigations_url(self) -> str:
+        """
+        Get request url for investigations result.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/investigations".format(self.base_url)
+
+    def _edr_historical_searches_result_url(self, search_id: str) -> str:
+        """
+        Get request url for historical searches result.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/searches/historical/{1}/results".format(self.base_url, search_id)
+
+    @property
+    def _edr_realtime_searches_query_url(self) -> str:
+        """
+        Get request url for historical searches query.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/searches/realtime".format(self.base_url)
+
+    def _edr_realtime_searches_result_url(self, search_id: str) -> str:
+        """
+        Get request url for historical searches result.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/searches/realtime/{1}/results".format(self.base_url, search_id)
+
+    def _edr_searches_job_status_url(self, search_id: str) -> str:
+        """
+        Get request url for historical searches job status.
+
+        Will return one of:
+            Status code 200 — The search is still in progress. Retry after 15 seconds.
+            Status code 303 — The search is finished. You can proceed to get results.
+
+        Delay between requests should be 15 seconds.
+
+        Returns:
+            str:
+        """
+        return "{0}/edr/v2/searches/queue-jobs/{1}".format(self.base_url, search_id)
+
+    async def _get_token_refresher(self, scopes: Set[Scope]) -> TrellixTokenRefresher:
+        """
+        Get token refresher.
+
+        Args:
+            scopes: Set[Scope]
+
+        Returns:
+            TrellixTokenRefresher:
+        """
+        return await TrellixTokenRefresher.instance(
+            self.client_id, self.client_secret, self.api_key, self.auth_url, scopes
+        )
+
+    async def _wait_for_search(self, search_id: str, scopes: Set[Scope]) -> None:
+        """
+        Wait for search to be finished.
+
+        Args:
+            search_id: str
+            scopes: Set[Scope]
+        """
+        search_status_url = self._edr_searches_job_status_url(search_id)
+        while True:
+            headers = self._request_headers(scopes)
+            async with self.session().get(search_status_url, headers=headers) as status_response:
+                if status_response.status == 303:
+                    break
+
+                # In documentation it is said that we should wait 15 seconds before next request
+                await asyncio.sleep(15)
```

Comment:
Could you limit the number of attempts to avoid infinite loop if the api isn't able to return a result?
The number of attempt could be pass as an argument, and you can replace the `while True:` by `for _ in range(max_attempts):`.

Reply:
The problem with this solution is that we do not know how much time will take to process this task.
I mean the flow is the next -
1. Create job
2. Wait until job change status to success/fail. It means the response status code will be 303
And to be honest I do now know how much time it will take. I found better solution with 
`aiolimiter.AsyncLimiter` instead of `asyncio.sleep`
https://aiolimiter.readthedocs.io/en/latest/
And implemented it is Salesforce integration, so will use it here also.

But still it will not give us a solution of avoiding while True statement as for me :(